pandas/core/internals/blocks.py

```diff
@@ -739,7 +739,10 @@ def to_native_types(self, slicer=None, na_rep='nan', quoting=None,
         mask = isna(values)
 
         if not self.is_object and not quoting:
-            values = values.astype(str)
+            if na_rep and isinstance(na_rep, str):
```

Comment:
this is not going to work, see how we do this in ``pandas/io/pytables`` in ``_convert_string_array``. we will want to adopt that approach.

Reply:
Tried reusing `_convert_string_array` first to see if that would work, but it returns `S{...}`, which causes tests to fail (e.g. expected `'mynull'`, instead got `b'mynull'`).

Then tried to re-use parts of the function. More specifically the part that calculates the text length (`bytes` or `string`). I created a function to be used in both `_convert_string_array` and by other callers.

Alas, I couldn't get it to work as a `cdef` (I believe Python Python cannot call it?), nor as `cpdef` (which was funny, but some tests were failing with being aborted with `SIGABRT` :confused: ). In the end, I only managed to get it to work by having a simple function, without type hinting.

@jreback probably not what you had in mind. Care to take a look and suggest an alternative whenever you have time, please? Pushed in a separate commit.

Cheers
Bruno